calico_node/readiness/readiness.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2018 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"flag"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/calico/calico_node/readiness/bird"
+)
+
+const felixReadinessEp = "http://localhost:9099/readiness"
+
+var enableBirdChecks = flag.Bool("bird", false, "Specify to enable bird readiness checks")
+var enableBird6Checks = flag.Bool("bird6", false, "Specify to enable bird6 readiness checks")
+var enableFelixChecks = flag.Bool("felix", false, "Specify to enable felix readiness checks")
+
+func main() {
+	flag.Parse()
+
+	if !*enableBirdChecks && !*enableFelixChecks && !*enableBird6Checks {
+		fmt.Printf("calico/node readiness check error: must specify at least one of -bird or -felix")
+		os.Exit(1)
+	}
+
+	if *enableFelixChecks {
+		if err := checkFelixReady(); err != nil {
+			fmt.Printf("calico/node is not ready: felix is not ready: %+v", err)
+			os.Exit(1)
+		}
+	}
+
+	if *enableBirdChecks {
+		if err := checkBirdReady("4"); err != nil {
+			fmt.Printf("calico/node is not ready: bird is not ready: %+v", err)
+			os.Exit(1)
+		}
+	}
+
+	if *enableBird6Checks {
+		if err := checkBirdReady("6"); err != nil {
+			fmt.Printf("calico/node is not ready: bird is not ready: %+v", err)
+			os.Exit(1)
+		}
+	}
+}
+
+// checkBirdReady checks if bird is ready by connecting to the bird
+// socket to gather all BGP peer connection status, and overall graceful
+// restart status.
+func checkBirdReady(ipv string) error {
+	// Check for unestablished peers
+	ueps, err := bird.GetBirdPeers(ipv)
+	log.Debugf("ueps: %v", ueps)
+	if err != nil {
+		return err
+	}
+
+	s := []string{}
+	for _, uep := range ueps {
+		if uep.BGPState != "Established" {
+			s = append(s, uep.PeerIP)
+		}
+	}
+	if len(s) > 0 {
+		return fmt.Errorf("BGP not established with %+v", strings.Join(s, ","))
+	}
+
+	// Check for GR
+	err = bird.CheckGrInProgress(ipv)
```

Comment:
I know we discussed a bit yesterday, but looking at this now I think it would be nicer if it was like so:

```go
gr, err := bird.GRInProgress(ipv)
if err != nil {
  return err
} else if gr {
  return errors.New("graceful restart in progress")
}
```